data_preparation/AgML_predictor_data_prepare.py

```diff
@@ -0,0 +1,1077 @@
+            max_extent_window_bounds = dataset.window_bounds(ds_window)
+            min_res_window_size = rasterio.windows.shape(ds_window)
+
+    # new bounds fitted to the highest resolution file
+    out_bounds = max_res_ds.window_bounds(round_window(max_res_ds.window(*max_extent_window_bounds)))
+    out_shape = rasterio.windows.shape(round_window(max_res_ds.window(*out_bounds)))
+
+    return out_bounds, out_shape
+
+"""
+@author: Joint Research Center
+"""
+def arr_stats(arr, weights=None,
+              output=('min', 'max', 'sum', 'mean', 'count', 'mode')):
+    """
+    Extracts statistics from input array (arr).
+    It uses weights array as weights if provided.
+    List of statistics to extract is defined in the output parameter as a list or a comma separated string.
+    :param arr: Input array
+    :param weights: Array providing weights for each pixel value. Used to calculate stats.
+    :param output: List of values to extract, can be a list or comma separated string.
+        Possible values are:
+            - min
+            - max
+            - sum - sum of all un masked arr pixels, no weight applied
+            - mean - average value, if weights are provided calculates a weighted average
+            - std - standard deviation, if weights are provided uses weighted calculation
+            - median - median values, WARNING doesn't use weights
+            - count - number of pixels used in a group, can be different to total number if a mask is applied
+            - weight_sum - sum of weights
+    :return: dict with calculated stats values
+    """
+    # prepare output and make sure it is a list
+    _output = output if type(output) in [list, tuple] else output.split()
+    _known_outputs = ('min', 'max', 'sum', 'mean', 'std', 'median', 'count', 'mode', 'weight_sum')  # todo handle more extractions
+    if not any(elem in _output for elem in _known_outputs):
+        raise Exception('Output not defined properly, should define at least one known output %s' % str(_known_outputs))
+
+    out_vals = dict()
+    # make sure array is a masked array
+    _arr = np.ma.array(arr)
+
+    if any(elem in _output for elem in ('std', 'min', 'max', 'sum', 'median', 'mode')):
+        arr_compressed = _arr.compressed()
+
+    if 'mean' in _output:
+        if weights is not None:
+            ind = np.isnan(_arr) | np.isnan(weights) | (_arr <= -9999)
+            out_vals['mean'] = np.ma.average(_arr[~ind], weights=weights[~ind])
+        else:
+            ind = np.isnan(_arr)
+            out_vals['mean'] = np.ma.mean(_arr[~ind])
+
+    if 'std' in _output:
+        arr_size = np.size(arr_compressed)
+        if weights is not None:
+            # combine mask from the arr and create a compressed arr
+            weights_compressed = np.ma.array(weights, mask=_arr.mask).compressed()
+            ind = np.isnan(arr_compressed) | np.isnan(weights_compressed) | (_arr <= -9999)
```

Comment:
These arrays might not be of the same shape, you are mixing compressed and uncompressed arrays